consensus/state.go

```diff
@@ -1463,6 +1471,16 @@ func (cs *State) finalizeCommit(height int64) {
 
 	fail.Fail() // XXX
 
+	// Prune old heights, if requested by ABCI app.
+	if retainHeight > 0 {
```

Comment:
Why not do all this inside ApplyBlock?

Reply:
Block application and block pruning seem orthogonal? I.e. it should be possible to apply a block without pruning anything.

Reply:
A related point here is that we currently only prune blocks during consensus, not during block replay. We may want to do both?

Reply:
They do seem orthogonal but given pruning is controlled by ABCI return values and all ABCI handling is supposed to be covered by ApplyBlock, I think it makes sense to consolidate.

Reply:
Opened #5414 for this, will pick it up once I wrap up some other projects.